// flatfile-csv-importer/build/main/results.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var stats_1 = require("./stats");
var user_1 = require("./user");
var upload_file_1 = require("./upload-file");
var streamed_results_1 = require("./streamed-results");
var FlatfileResults = /** @class */ (function () {
    function FlatfileResults(data, meta, importer) {
        this.$meta = meta;
        this.$data = data;
        this.$importer = importer;
    }
    Object.defineProperty(FlatfileResults.prototype, "rawOutput", {
        /**
         * The raw output from the importer including all deleted rows
         * and sequence info
         */
        get: function () {
            return this.blobOnly(this.$data, 'rawOutput');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "data", {
        /**
         * An array of valid data, key-mapped to the configuration provided
         * (alias of validData)
         */
        get: function () {
            return this.blobOnly(this.validData, 'data');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "validData", {
        /**
         * An array of valid data, key-mapped to the configuration provided
         */
        get: function () {
            var res = this.$data.filter(function (v) { return v.valid; })
                .filter(function (v) { return !v.deleted; })
                .map(function (v) { return v.data; });
            return this.blobOnly(res, 'validData');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "deletedData", {
        /**
         * Rows of data the user excluded from the final results,
         * key-mapped to the configuration provided
         */
        get: function () {
            var res = this.$data.filter(function (v) { return v.deleted; })
                .map(function (v) { return v.data; });
            return this.blobOnly(res, 'deletedData');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "allData", {
        /**
         * All data from the original file upload including deleted rows,
         * key-mapped to the configuration provided
         */
        get: function () {
            return this.blobOnly(this.$data.map(function (v) { return v.data; }), 'allData');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "batchId", {
        /**
         * The uuid of the batch assigned by Flatfile (use this in internal
         * references for support purposes)
         */
        get: function () {
            return this.$meta.batchID;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "stats", {
        /**
         * Stats and counts about this file upload
         */
        get: function () {
            return new stats_1.default(this.$meta);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "customer", {
        /**
         * The customer provided in setCustomer
         */
        get: function () {
            if (this.$meta.endUser) {
                return new user_1.default(this.$meta.endUser);
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "originalFile", {
        /**
         * A File object of the originally uploaded file stored as an AWS url
         */
        get: function () {
            if (this.$meta.originalFile) {
                return new upload_file_1.default(this.$meta.originalFile);
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "csvFile", {
        /**
         * Same as originalFile unless it was uploaded in xls format, in which case this is the converted csv file stored as an AWS url
         */
        get: function () {
            if (this.$meta.originalFile) {
                if (this.$meta.originalFile.filetype === 'csv') {
                    return new upload_file_1.default(this.$meta.originalFile);
                }
                else {
                    if (this.$meta.csvFile) {
                        return new upload_file_1.default(this.$meta.csvFile);
                    }
                }
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "fileName", {
        /**
         * The filename of the originally uploaded file
         */
        get: function () {
            return this.$meta.filename || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "managed", {
        /**
         * If the final upload is managed by a private endpoint or not
         */
        get: function () {
            return this.$meta.managed || false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "manual", {
        /**
         * If the data was entered manually instead of via file upload or not
         */
        get: function () {
            return this.$meta.manual;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "config", {
        /**
         * The parsed and bootstrapped config object used by this importer instance
         */
        get: function () {
            return this.$meta.config;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "parsingConfig", {
        /**
         * The configuration used by the csv parser PapaParse: https://www.papaparse.com/docs#config
         */
        get: function () {
            return this.$meta.parsing_config;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "skippedRows", {
        /**
         * The invalid rows that were skipped on submission
         */
        get: function () {
            return this.$meta.skipped_rows || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "headersRaw", {
        /**
         * The headers before they were matched as given in the original file
         */
        get: function () {
            return this.$meta.headers_raw || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "headersMatched", {
        /**
         * The headers after they are matched
         */
        get: function () {
            return this.$meta.headers_matched || null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Get the next chunk of records
     */
    FlatfileResults.prototype.nextChunk = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            if (!_this.$meta.inChunks) {
                return reject("\"nextChunk()\" is only accessible when using \"inChunks\". Please see docs for \"requestDataFromUser\".");
            }
            _this.$importer.$ready.then(function (child) {
                console.log('child.nextChunk()');
                child.nextChunk().then(function (data) {
                    console.log('nextChunk()', data);
                    resolve(data.results.length ? new streamed_results_1.default(data.results, data.meta) : null);
                }, function (err) {
                    console.log('nextChunk(err)', err);
                });
            });
        });
    };
    Object.defineProperty(FlatfileResults.prototype, "customColumns", {
        /**
         * An array of any columns that were created during import
         */
        get: function () {
            return this.$meta.custom_columns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "failureReason", {
        /**
         * The reason for the failure if there was a failure
         */
        get: function () {
            return this.$meta.failure_reason || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "submittedAt", {
        /**
         * The time that the data was submitted
         */
        get: function () {
            return this.$meta.submitted_at || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "failedAt", {
        /**
         * The time that the import failed if it failed
         */
        get: function () {
            return this.$meta.failed_at || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlatfileResults.prototype, "createdAt", {
        /**
         * The time the data began the import, whether via file upload or manual data entry
         */
        get: function () {
            return this.$meta.created_at;
        },
        enumerable: true,
        configurable: true
    });
    FlatfileResults.prototype.blobOnly = function (v, method, alt) {
        if (alt === void 0) { alt = 'nextChunk()'; }
        if (this.$meta.inChunks) {
            throw new Error("\"" + method + "\" is not accessible when using \"inChunks\". Please see docs for \"" + alt + "\" instead.");
        }
        return v;
    };
    return FlatfileResults;
}());
exports.default = FlatfileResults;
